lib/autoscale/aws/events.go

```diff
@@ -148,7 +168,203 @@ func (a *Autoscaler) ensureInstanceTerminated(ctx context.Context, event HookEve
+
+		a.monitorShrink(ctx, operator, event, opKey)
+
+		return nil
+	})
+
+	if err != nil {
+		log.WithError(err).Warn("Autoscaler failed to shrink cluster in response to ASG event")
+	}
+}
+
+// monitorShrink monitors the cluster shrink operation and responds to failures of the shrink operation.
+func (a *Autoscaler) monitorShrink(
+	ctx context.Context,
+	operator Operator,
+	event HookEvent,
+	opKey *ops.SiteOperationKey) {
+	log := a.WithFields(logrus.Fields{
+		"instance": event.InstanceID,
+		"asg_name": event.AutoScalingGroupName,
+		"op_key":   opKey,
+	})
+
+	ctx, cancel := context.WithTimeout(ctx, eventTimeout)
+	defer cancel()
+
+	ticker := time.NewTicker(monitorShrinkInterval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Warn("Exiting ASG shrink monitor due to reaching max lifetime")
+			return
+		case <-ticker.C:
+			progress, err := operator.GetSiteOperationProgress(*opKey)
+			if err != nil {
+				log.WithError(err).Warn("failed to retrieve progress entry")
+				continue
+			}
+
+			if progress.IsCompleted() {
+				return
+			}
+
+			// if the shrink operation fails, let AWS shutdown the instance by completing the lifecycle event, and then
+			// try to run a force shrink on an offline node
+			if progress.State == ops.ProgressStateFailed {
+				err := a.completeASGLifecycle(event)
+				if err != nil {
+					// AWS might return errors if we try and complete an ASG lifecycle that is no longer valid due to
+					// timeout. So skip the error and continue with the force shrink.
+					log.WithError(err).Warn("failed to send AWS ASG completion event")
+				}
+
+				err = a.ensureInstanceTerminated(ctx, event)
+				if err != nil {
+					log.WithError(err).Warn("failed to ensure AWS instance has been terminated")
+					continue
+				}
+
+				err = a.forceShrink(ctx, operator, event)
+				if err != nil {
+					log.WithError(err).Warn("failed to force shrink operation")
+					continue
+				}
+
+				// Once we've started the force shrink operation, exit the monitoring loop. If the force shrink fails
+				// this loop isn't really capable of recovering and shouldn't just continually try to shrink the node.
+				return
+			}
+		}
+	}
+}
+
+func (a *Autoscaler) completeASGLifecycle(event HookEvent) error {
+	_, err := a.Config.AutoScaling.CompleteLifecycleAction(&autoscaling.CompleteLifecycleActionInput{
+		AutoScalingGroupName:  aws.String(event.AutoScalingGroupName),
+		InstanceId:            aws.String(event.InstanceID),
+		LifecycleActionToken:  aws.String(event.Token),
+		LifecycleHookName:     aws.String(event.LifecycleHookName),
+		LifecycleActionResult: aws.String("CONTINUE"),
+	})
+
+	a.WithError(err).WithFields(log.Fields{
+		"instance": event.InstanceID,
+		"asg_name": event.AutoScalingGroupName,
+	}).Info("notified AWS of completed uninstall")
+
+	return trace.Wrap(err)
+}
+
+func (a *Autoscaler) startHeartbeatLoop(event HookEvent) context.CancelFunc {
+	ctx, cancel := context.WithTimeout(context.Background(), eventTimeout)
```

Comment:
Should this inherit from the `RemoveInstance` context? It looks like you instead return the cancel fn and use it in a defer in the parent.  It seems like similar functionality -- is there a reason to prefer the latter here?

Reply:
Yea, you're right, the code doesn't quite make sense, as the parent context was indirectly tied to the heartbeat context because if the parent context is cancelled, the defer would cancel the heartbeat.

I'm not sure why I didn't originally tie the heartbeat to the parent getting cancelled, other than it's one of the context's that we use to signal service shutdown, which is a pattern used in the gravity code but doesn't actually tend to work as specified. 

We still want a separate context for the heartbeat, so that it times out around the same time as the node, and doesn't keep hitting the AWS API and respective rate limits if the heartbeat has no chance of progressing, but removeNode can still be running after that point.